Add doc for List partition and List columns partition (#5132)
pingcap/docs-cn: partitioned-table.md

## 分区类型

> 本节介绍在 TiDB 中的分区类型。当前支持的类型包括 [Range 分区](#range-分区)、[List 分区](#list-分区)和 [Hash 分区](#hash-分区)。Range 分区和 List 分区可以用于解决业务中大量删除带来的性能问题，支持快速删除分区。Hash 分区则可以用于大量写入场景下的数据打散。

Reviewer: ```suggestion
本节介绍在 TiDB 中的分区类型。当前支持的类型包括 [Range 分区](#range-分区)、[List 分区](#list-分区)、[List COLUMNS 分区](#list-columns-分区) 和 [Hash 分区](#hash-分区)。Range 分区和 List 分区可以用于解决业务中大量删除带来的性能问题，支持快速删除分区。Hash 分区则可以用于大量写入场景下的数据打散。
```

List 分区和 Range 分区有很多相似的地方，主要的不同在于，List 分区中，对于表的每个分区中包含的所有行，按分区表达式计算的值属于给定的数据集合内。每个分区定义的数据集合是任意个值，但不能有重叠，通过使用 `VALUES IN` 进行定义。

下列场景中，假设你要创建一个人事记录的表：

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
);
```

假如一共有 20 个商店分别属于 4 个地区中，如下表所示：

| Region  | Store ID Numbers     |
| ------- | -------------------- |
| North   | 3, 5, 6, 9, 17       |
| East    | 1, 2, 10, 11, 19, 20 |
| West    | 4, 12, 13, 14, 18    |
| Central | 7, 8, 15, 16         |

如果想把同一个地区的商店数据都存储在同一个分区里面，你可以根据 `store_id` 来创建 List 分区：

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT
)
PARTITION BY LIST(store_id) (
    PARTITION pNorth VALUES IN (3,5,6,9,17),
    PARTITION pEast VALUES IN (1,2,10,11,19,20),
    PARTITION pWest VALUES IN (4,12,13,14,18),
    PARTITION pCentral VALUES IN (7,8,15,16)
);
```

> 这样在表中添加或删除于特定区域相关的记录将变得非常容易。例如，假设西部地区所有的商店都卖给了另一家公司，所有与该地区商店的员工相关的行数据都可以通过 `ALTER TABLE employees TRUNCATE PARTITION pWest` 被删除，这比等效的 DELETE 语句 `DELETE FROM employees WHERE store_id IN (4,12,13,14,18)` 执行更加高效。（使用 `ALTER TABLE员工DROP PARTITION pWest` 也会删除所有这些行，但也会从表的定义中删除分区 `pWest`；您将需要使用 `ALTER TABLE ... ADD PARTITION` 语句来还原表的原始分区方案。）

Reviewer: ```suggestion
这样就能方便地在表中添加或删除与特定区域相关的记录。例如，假设西部地区 (West) 所有的商店都卖给了另一家公司，所有该地区商店的员工相关的行数据都可以通过 `ALTER TABLE employees TRUNCATE PARTITION pWest` 删除，这比等效的 `DELETE` 语句 `DELETE FROM employees WHERE store_id IN (4,12,13,14,18)` 执行起来更加高效。

使用 `ALTER TABLE employees DROP PARTITION pWest` 也能删除所有这些行，但同时也会从表的定义中删除分区 `pWest`。那样你还需要使用 `ALTER TABLE ... ADD PARTITION` 语句来还原表的原始分区方案。
```

> 与RANGE分区的情况不同，它没有诸如 `MAXVALUE` 之类的“包罗万象”的东西。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)`子句中。包含不匹配分区列值的 `INSERT` 语句将执行失败，并显示错误，如下例所示：

Reviewer: 什么是“包罗万象”呢？
```suggestion
与 Range 分区的情况不同，List 分区没有诸如 `MAXVALUE` 之类的“包罗万象”的东西。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)` 子句中。如果 `INSERT` 语句包含不匹配分区列值，该语句将执行失败并报错，如下例所示：
```

```sql
mysql> CREATE TABLE h2 (
    ->   c1 INT,
    ->   c2 INT
    -> )
    -> PARTITION BY LIST(c1) (
    ->   PARTITION p0 VALUES IN (1, 4, 7),
    ->   PARTITION p1 VALUES IN (2, 5, 8)
    -> );
Query OK, 0 rows affected (0.11 sec)

mysql> INSERT INTO h2 VALUES (3, 5);

> ERROR 1525 (HY000): Table has no partition for value 3

Reviewer: + 这个示例的数据可以改改吗？
+ 去掉 mysql> ? 这样会比较明显。

```

> 可以通过使用 `IGNORE` 关键字来忽略这种类型的错误。如果这样做，则不会插入包含不匹配分区列值的行，但是会插入任何具有匹配值的行，并且不会报告错误:

Reviewer: ```suggestion
要忽略以上类型的错误，可以通过使用 `IGNORE` 关键字。使用后，就不会插入包含不匹配分区列值的行，但是会插入任何具有匹配值的行，并且不会报错:
```

```sql
mysql> TRUNCATE h2;
Query OK, 1 row affected (0.00 sec)

mysql> SELECT * FROM h2;
Empty set (0.00 sec)

mysql> INSERT IGNORE INTO h2 VALUES (2, 5), (6, 10), (7, 5), (3, 1), (1, 9);
Query OK, 3 rows affected (0.00 sec)
Records: 5  Duplicates: 2  Warnings: 0

mysql> SELECT * FROM h2;
+------+------+
| c1   | c2   |
+------+------+
|    7 |    5 |
|    1 |    9 |
|    2 |    5 |
+------+------+
3 rows in set (0.00 sec)
```

### List COLUMNS 分区

> LIST COLUMNS 分区是 LIST 分区的一种变体，它可以将多个列用作分区键，并且可以将整数类型以外的数据类型的列用作分区列；您可以使用字符串类型，DATE 和 DATETIME 类型的列。

Reviewer: ```suggestion
LIST COLUMNS 分区是 List 分区的一种变体，可以将多个列用作分区键，并且可以将整数类型以外的数据类型的列用作分区列。你还可以使用字符串类型、`DATE` 和 `DATETIME` 类型的列。
```

假设您有一家在12个城市拥有客户的企业，出于销售和营销目的，您将组织成3个城市的4个区域，如下表所示：

| Region | Cities                         |
| :----- | ------------------------------ |
| 1      | Oskarshamn, Högsby, Mönsterås  |
| 2      | Vimmerby, Hultsfred, Västervik |
| 3      | Nässjö, Eksjö, Vetlanda        |

> | 4      | Uppvidinge, Alvesta, Växjo     |

Reviewer: 这些案例能否换成美国地名呢？

Author: 这里是和 MySQL 的文档用的相同的例子 (⊙o⊙)… 感觉还是和 MySQL 文档保持一致吧？

Reviewer: 这是我们的文档。要是连例子都一样了，岂不有抄袭的嫌疑了？

Author: em....  那我改下测试用例吧，组织结构感觉不好改

使用列表列分区，您可以为客户数据创建一个表，每行数据将根据客户所在城市的名称分配给与这些区域对应的4个分区中的任意一个，如下所示：

{{< copyable "sql" >}}

```sql
CREATE TABLE customers_1 (
    first_name VARCHAR(25),
    last_name VARCHAR(25),
    street_1 VARCHAR(30),
    street_2 VARCHAR(30),
    city VARCHAR(15),
    renewal DATE
)
PARTITION BY LIST COLUMNS(city) (
    PARTITION pRegion_1 VALUES IN('Oskarshamn', 'Högsby', 'Mönsterås'),
    PARTITION pRegion_2 VALUES IN('Vimmerby', 'Hultsfred', 'Västervik'),
    PARTITION pRegion_3 VALUES IN('Nässjö', 'Eksjö', 'Vetlanda'),

>     PARTITION pRegion_4 VALUES IN('Uppvidinge', 'Alvesta', 'Växjo')

Reviewer: 同上

);

> ```

Reviewer: 基本是照抄 MySQL 的文档了，可能有侵权的风险吧。

与 List 分区不同的是，您不需要在 `COLUMNS()` 子句中使用表达式来将列值转换为整数。

> LIST COLUMNS 分区也可以使用 DATE 和 DATETIME 类型的列进行分区，如以下示例中所示，该示例使用与先前显示的 `customer_1` 表相同的名称和列，但根据 `renewal` 列采用 LIST COLUMNS 分区将行存储在4个分区之一中：

Reviewer: ```suggestion
LIST COLUMNS 分区也可以使用 `DATE` 和 `DATETIME` 类型的列进行分区，如以下示例所示。该示例使用与先前显示的 `customer_1` 表相同的名称和列，但根据 `renewal` 列采用 LIST COLUMNS 分区将行存储在 4 个分区之一中：
```

{{< copyable "sql" >}}

```sql
CREATE TABLE customers_2 (
    first_name VARCHAR(25),
    last_name VARCHAR(25),
    street_1 VARCHAR(30),
    street_2 VARCHAR(30),
    city VARCHAR(15),
    renewal DATE
)
PARTITION BY LIST COLUMNS(renewal) (
    PARTITION pWeek_1 VALUES IN('2010-02-01', '2010-02-02', '2010-02-03',
        '2010-02-04', '2010-02-05', '2010-02-06', '2010-02-07'),
    PARTITION pWeek_2 VALUES IN('2010-02-08', '2010-02-09', '2010-02-10',
        '2010-02-11', '2010-02-12', '2010-02-13', '2010-02-14'),
    PARTITION pWeek_3 VALUES IN('2010-02-15', '2010-02-16', '2010-02-17',
        '2010-02-18', '2010-02-19', '2010-02-20', '2010-02-21'),
    PARTITION pWeek_4 VALUES IN('2010-02-22', '2010-02-23', '2010-02-24',
        '2010-02-25', '2010-02-26', '2010-02-27', '2010-02-28')
);

> ```

Reviewer: 用例雷同，有侵权风险。

> 另外，您也可以在 `COLUMNS()` 子句中使用多个列，例如：

Reviewer: ```suggestion
另外，你也可以在 `COLUMNS()` 子句中添加多个列，例如：
```

* 使用包含时间或者日期的列，或者是其它按序生成的数据。
* 频繁查询分区使用的列。例如执行这样的查询 `EXPLAIN SELECT COUNT(*) FROM employees WHERE separated BETWEEN '2000-01-01' AND '2000-12-31' GROUP BY store_id;` 时，TiDB 可以迅速确定，只需要扫描 `p2` 分区的数据，因为其它的分区不满足 `where` 条件。

### List 分区

> **警告：**
>
> 该功能目前为实验特性，不建议在生产环境中使用。

在创建 List 分区表之前，需要先将 session 变量 `tidb_enable_table_partition` 的值设置为 `ON`。

{{< copyable "sql" >}}

```sql

> set @@session.tidb_enable_table_partition = 'ON';

Reviewer: ```suggestion
set @@session.tidb_enable_table_partition = 'ON';
```
```

* 使用包含时间或者日期的列，或者是其它按序生成的数据。
* 频繁查询分区使用的列。例如执行这样的查询 `EXPLAIN SELECT COUNT(*) FROM employees WHERE separated BETWEEN '2000-01-01' AND '2000-12-31' GROUP BY store_id;` 时，TiDB 可以迅速确定，只需要扫描 `p2` 分区的数据，因为其它的分区不满足 `where` 条件。

### List 分区

> **注意：**
>
> 该功能目前为实验特性，不建议在生产环境中使用。

如果需要使用这一特性，可将 session 变量 `tidb_enable_table_partition` 的值设置为 `ON` 后，再创建 List 分区表。

{{< copyable "sql" >}}

```sql
set @@session.tidb_enable_table_partition = 'ON';
```

> List 分区和 Range 分区有很多相似的地方，主要的不同在于 List 分区中，对于表的每个分区中包含的所有行，按分区表达式计算的值属于给定的数据集合。每个分区定义的数据集合有任意个值，但值不能有重叠，可通过 `PARTITION ... VALUES IN (...)` 子句进行定义值。

Reviewer: ```suggestion
List 分区和 Range 分区有很多相似的地方。不同之处主要在于 List 分区中，对于表的每个分区中包含的所有行，按分区表达式计算的值属于给定的数据集合。每个分区定义的数据集合有任意个值，但不能有重复的值，可通过 `PARTITION ... VALUES IN (...)` 子句对值进行定义。
```

假设你要创建一张人事记录表，示例如下：

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
);
```

假如一共有 20 个商店分布在 4 个地区，如下表所示：

| Region  | Store ID Numbers     |
| ------- | -------------------- |
| North   | 1, 2, 3, 4, 5        |
| East    | 6, 7, 8, 9, 10       |
| West    | 11, 12, 13, 14, 15   |
| Central | 16, 17, 18, 19, 20   |

> 如果想把同一个地区的商店的员工人事数据都存储在同一个分区里面，你可以根据 `store_id` 来创建 List 分区：

Reviewer: ```suggestion
如果想把同一个地区商店员工的人事数据都存储在同一个分区中，你可以根据 `store_id` 来创建 List 分区：
```

{{< copyable "sql" >}}

```sql
CREATE TABLE employees (
    id INT NOT NULL,
    hired DATE NOT NULL DEFAULT '1970-01-01',
    store_id INT
)
PARTITION BY LIST (store_id) (
    PARTITION pNorth VALUES IN (1, 2, 3, 4, 5),
    PARTITION pEast VALUES IN (6, 7, 8, 9, 10),
    PARTITION pWest VALUES IN (11, 12, 13, 14, 15),
    PARTITION pCentral VALUES IN (16, 17, 18, 19, 20)
);
```

> 这样就能方便地在表中添加或删除与特定区域相关的记录。例如，假设东部地区 (East) 所有的商店都卖给了另一家公司，所有该地区商店的员工相关的行数据都可以通过 `ALTER TABLE employees TRUNCATE PARTITION pEast` 删除，这比等效的 `DELETE` 语句 `DELETE FROM employees WHERE store_id IN (6, 7, 8, 9, 10)` 执行起来更加高效。

Reviewer: ```suggestion
这样就能方便地在表中添加或删除与特定区域相关的记录。例如，假设东部地区 (East) 所有的商店都卖给了另一家公司，所有该地区商店员工相关的行数据都可以通过 `ALTER TABLE employees TRUNCATE PARTITION pEast` 删除，这比等效的 `DELETE` 语句 `DELETE FROM employees WHERE store_id IN (6, 7, 8, 9, 10)` 执行起来更加高效。
```

使用 `ALTER TABLE employees DROP PARTITION pEast` 也能删除所有这些行，但同时也会从表的定义中删除分区 `pEast`。那样你还需要使用 `ALTER TABLE ... ADD PARTITION` 语句来还原表的原始分区方案。

> 与 Range 分区的情况不同，List 分区没有诸如 `MAXVALUE` 之类的“包罗万象”的东西。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)` 子句中。如果 `INSERT` 语句包含不匹配分区列值，该语句将执行失败并报错，如下例所示：

Reviewer: ```suggestion
与 Range 分区的情况不同，List 分区没有诸如 `MAXVALUE` 之类的“包罗万象”的东西。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)` 子句中。如果 `INSERT` 语句要插入的值不匹配分区的列值，该语句将执行失败并报错，如下例所示：
```

> 与 Range 分区的情况不同，List 分区没有诸如 `MAXVALUE` 之类的“包罗万象”的东西。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)` 子句中。如果 `INSERT` 语句包含不匹配分区列值，该语句将执行失败并报错，如下例所示：

Author: Range partition 可以最后定义 less than maxValue 的 partition, 让 一个 value 不属于某个 partition 时，就放在最后的 maxValue partition 里面。但是 list partition 没有类似这种定义的 partition。

```sql
test> CREATE TABLE t (
    ->   a INT,
    ->   b INT
    -> )
    -> PARTITION BY LIST (a) (
    ->   PARTITION p0 VALUES IN (1, 2, 3),
    ->   PARTITION p1 VALUES IN (4, 5, 6)
    -> );
Query OK, 0 rows affected (0.11 sec)

test> INSERT INTO t VALUES (7, 7);
ERROR 1525 (HY000): Table has no partition for value 7
```

> 要忽略以上类型的错误，可以通过使用 `IGNORE` 关键字。使用后，就不会插入包含不匹配分区列值的行，但是会插入任何具有匹配值的行，并且不会报错:

Reviewer: ```suggestion
要忽略以上类型的错误，可以通过使用 `IGNORE` 关键字。使用该关键字后，就不会插入包含不匹配分区列值的行，但是会插入任何具有匹配值的行，并且不会报错:
```

```sql
test> TRUNCATE t;
Query OK, 1 row affected (0.00 sec)

test> INSERT IGNORE INTO t VALUES (1, 1), (7, 7), (8, 8), (3, 3), (5, 5);
Query OK, 3 rows affected, 2 warnings (0.01 sec)
Records: 5  Duplicates: 2  Warnings: 2

test> select * from t;
+------+------+
| a    | b    |
+------+------+
|    5 |    5 |
|    1 |    1 |
|    3 |    3 |
+------+------+
3 rows in set (0.01 sec)
```

### List COLUMNS 分区

LIST COLUMNS 分区是 List 分区的一种变体，可以将多个列用作分区键，并且可以将整数类型以外的数据类型的列用作分区列。你还可以使用字符串类型、`DATE` 和 `DATETIME` 类型的列。

假设您的员工分别来自以下 12 个城市，想要根据相关规定分成 4 个区域，如下表所示：

| Region | Cities                         |
| :----- | ------------------------------ |
| 1      | LosAngeles,Seattle, Houston    |
| 2      | Chicago, Columbus, Boston      |
| 3      | NewYork, LongIsland, Baltimore |
| 4      | Atlanta, Raleigh, Cincinnati   |

> 使用列表列分区，您可以为员工数据创建一个表，每行数据将根据员工所在城市名称与这些区域对应的4个分区中的任意一个，如下所示：

Author: 已改成 “使用列表列分区，你可以为员工数据创建一个表，将每行数据存储在员工所在城市对应的分区中，如下所示：”

{{< copyable "sql" >}}

```sql
CREATE TABLE employees_1 (
    id INT NOT NULL,
    fname VARCHAR(30),
    lname VARCHAR(30),
    hired DATE NOT NULL DEFAULT '1970-01-01',
    separated DATE NOT NULL DEFAULT '9999-12-31',
    job_code INT,
    store_id INT,
    city VARCHAR(15)
)
PARTITION BY LIST COLUMNS(city) (
    PARTITION pRegion_1 VALUES IN('LosAngeles', 'Seattle', 'Houston'),
    PARTITION pRegion_2 VALUES IN('Chicago', 'Columbus', 'Boston'),
    PARTITION pRegion_3 VALUES IN('NewYork', 'LongIsland', 'Baltimore'),
    PARTITION pRegion_4 VALUES IN('Atlanta', 'Raleigh', 'Cincinnati')
);
```

> 与 List 分区不同的是，您不需要在 `COLUMNS()` 子句中使用表达式来将列值转换为整数。

Reviewer: ```suggestion
与 List 分区不同的是，你不需要在 `COLUMNS()` 子句中使用表达式来将列值转换为整数。
```

使用 `ALTER TABLE employees DROP PARTITION pEast` 也能删除所有这些行，但同时也会从表的定义中删除分区 `pEast`。那样你还需要使用 `ALTER TABLE ... ADD PARTITION` 语句来还原表的原始分区方案。

与 Range 分区的情况不同，List 分区没有类似的 `MAXVALUE` 分区来存储所有不属于其他 partition 的值。分区表达式的所有期望值都应包含在 `PARTITION ... VALUES IN (...)` 子句中。如果 `INSERT` 语句要插入的值不匹配分区的列值，该语句将执行失败并报错，如下例所示：

```sql
test> CREATE TABLE t (
    ->   a INT,
    ->   b INT
    -> )
    -> PARTITION BY LIST (a) (
    ->   PARTITION p0 VALUES IN (1, 2, 3),
    ->   PARTITION p1 VALUES IN (4, 5, 6)
    -> );
Query OK, 0 rows affected (0.11 sec)

test> INSERT INTO t VALUES (7, 7);
ERROR 1525 (HY000): Table has no partition for value 7
```

要忽略以上类型的错误，可以通过使用 `IGNORE` 关键字。使用该关键字后，就不会插入包含不匹配分区列值的行，但是会插入任何具有匹配值的行，并且不会报错:

```sql
test> TRUNCATE t;
Query OK, 1 row affected (0.00 sec)

test> INSERT IGNORE INTO t VALUES (1, 1), (7, 7), (8, 8), (3, 3), (5, 5);
Query OK, 3 rows affected, 2 warnings (0.01 sec)
Records: 5  Duplicates: 2  Warnings: 2

test> select * from t;
+------+------+
| a    | b    |
+------+------+
|    5 |    5 |
|    1 |    1 |
|    3 |    3 |
+------+------+
3 rows in set (0.01 sec)
```

### List COLUMNS 分区

> 

Reviewer: ```suggestion

> **警告：**
>
> 该功能从 v5.0.0-rc 引入，目前为实验特性，不建议在生产环境中使用。

```

### 兼容性

> 目前 TiDB 里面只实现了 Range 分区，List 分区和 Hash 分区，其它的 MySQL 分区类型比如 Key 分区尚不支持。

Reviewer: ```suggestion
目前 TiDB 里面只实现了 Range 分区、List 分区和 Hash 分区，其它的 MySQL 分区类型比如 Key 分区尚不支持。
```

### 兼容性

> 目前 TiDB 里面只实现了 Range 分区、List 分区和 Hash 分区，其它的 MySQL 分区类型比如 Key 分区尚不支持。

Reviewer: ```suggestion
目前 TiDB 里面只实现了 Range 分区、List 分区、List COLUMNS 和 Hash 分区，其它的 MySQL 分区类型比如 Key 分区尚不支持。
```

> 目前 TiDB 里面只实现了 Range 分区、List 分区、List COLUMNS 和 Hash 分区，其它的 MySQL 分区类型例如 Key 分区尚不支持。

Reviewer: ```suggestion
目前 TiDB 里面只实现了 Range 分区、List 分区、List COLUMNS 分区和 Hash 分区，其它的 MySQL 分区类型例如 Key 分区尚不支持。
```
